scripts/omni_run.py

```diff
@@ -2,220 +2,184 @@
+    # get correct firstcal files
+    # XXX not a fan of the way this is done, open to suggestions
+    fcalfile = None
+    if len(pols) == 1: #single pol
+        fcalfile = file2firstcal[file_group[pols[0]]]
+    else:
+        for pp in pols: #4 pol
+            if pp not in linear_pol_keys:
+                fcalfile = file2firstcal[file_group[pp]]
+                break
+    if not fcalfile: #2 pol
+        fcalfile = [file2firstcal[file_group[pp]][0] for pp in linear_pol_keys]
+       
+    _, g0, _, _ = from_fits(fcalfile)
+    
+    #uvd = pyuvdata.UVData()
+    #uvd.read_miriad([file_group[pp] for pp in pols])
+    #XXX THIS WILL BECOME MUCH SIMPLER WHEN PYUVDATA CAN READ MULTIPLE MIRIAD FILES AT ONCE
+    
+    ## collect metadata -- should be the same for each file
+    f0 = file_group[pols[0]]
     uvd = pyuvdata.UVData()
-    uvd.read_miriad(file_group[file_pol])
+    uvd.read_miriad(f0)
     t_jd = uvd.time_array.reshape(uvd.Ntimes, uvd.Nbls)[:,0]
     t_lst = uvd.lst_array.reshape(uvd.Ntimes, uvd.Nbls)[:,0]
+    t_int = uvd.integration_time
     freqs = uvd.freq_array[0]
     SH = (uvd.Ntimes, uvd.Nfreqs)  # shape of file data (ex: (19,203))
-    d, f = {}, {}
+    
+    uvd_dict = {}
+    for pp in pols:
+        uvd = pyuvdata.UVData()
+        uvd.read_miriad(file_group[pp])
+        uvd_dict[pp] = uvd
+    
+    ## format g0 for application to data
     for p in g0.keys():
         for i in g0[p]:
-            if g0[p][i].shape != (len(t_jd), len(freqs)):
+            if g0[p][i].shape != (len(t_jd), len(freqs)): #not a big fan of this if/else
                 g0[p][i] = np.resize(g0[p][i], SH)  # resize gains like data
             else: continue
-
-    for ip, pol in enumerate(uvd.polarization_array):
-        pol = aipy.miriad.pol2str[pol]
-        if pol != opts.pol: 
-            continue
-        for nbl, (i,j) in enumerate(map(uvd.baseline_to_antnums, uvd.baseline_array[:uvd.Nbls])):
-            if not (i, j) in bls and not (j, i) in bls:
-                continue
-            d[(i,j)] = {pol: uvd.data_array.reshape(uvd.Ntimes, uvd.Nbls, uvd.Nspws, uvd.Nfreqs, uvd.Npols)[:, nbl, 0, :, ip]}
-            f[(i,j)] = {pol: np.logical_not(uvd.flag_array.reshape(uvd.Ntimes, uvd.Nbls, uvd.Nspws, uvd.Nfreqs, uvd.Npols)[:, nbl, 0, :, ip])}
     
+    ## read data into dictionaries
+    d,f = {},{}
+    for ip,pp in enumerate(pols):
+        uvdp = uvd_dict[pp]
+        
+        if ip==0:
+            for nbl, (i,j) in enumerate(map(uvdp.baseline_to_antnums, uvdp.baseline_array[:uvdp.Nbls])):
+                if not (i, j) in bls and not (j, i) in bls: continue
+                d[(i,j)] = {}
+                f[(i,j)] = {}
+        
+        #XXX I *really* don't like looping again, but I'm not sure how better to do it
+        for nbl, (i,j) in enumerate(map(uvdp.baseline_to_antnums, uvdp.baseline_array[:uvdp.Nbls])):
+            if not (i, j) in bls and not (j, i) in bls: continue
```

Comment:
I think the with-parentheses version is easier to interpret as the baseline that it is. Otherwise I find it harder to keep track of the various indices for debugging etc.